pkg/ingester/bucket.go

```diff
@@ -2,67 +2,238 @@ package ingester
 
 import (
 	"fmt"
+	"hash/fnv"
 	"io"
+	"path/filepath"
+	"sort"
 	"strings"
+	"sync"
+	"time"
 
+	b3 "github.com/cortexproject/cortex/pkg/storage/tsdb/backend/s3"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block"
 	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/objstore/s3"
 	"golang.org/x/net/context"
 )
 
-// Bucket is a wrapper around a objstore.Bucket that prepends writes with a userID
+// Bucket is a wrapper around a slice of objstore.Bucket that prepends writes with a userID and shards them based on root name
 type Bucket struct {
```

Comment:
From a design perspective, this looks weird to me. A `Bucket` struct contains multiple `PeriodBucket`. Naming wise, to me a `Bucket` is a single bucket and then you have a storage abstraction which - under the hood is composed by multiple buckets.

I would suggest to rework a bit on this. An option may be keeping the  `Bucket` as is in master (all in all, a bucket is a bucket) and then add an abstraction over it, which incapsulate a "multiple buckets storage".

Following up, this way we may also keep the current implementation as is and opt-in into the multi-buckets only if multiple endpoints are configured. The idea is to keep two backend storages implementation: one single bucket and one multi-bucket. The multi-bucket is based on the single-bucket, for the single bucket implementation.